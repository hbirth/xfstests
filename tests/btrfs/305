#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2023 Western Digital Cooperation.  All Rights Reserved.
#
# FS QA Test 305
#
# Test on-disk layout of RAID Stripe Tree Metadata by writing 8k to a new file
# with a filesystem prepropulated, so that 4k of the write are written to the
# 1st stripe and 4k start a new stripe.
#
. ./common/preamble
_begin_fstest auto quick raid remount volume raid-stripe-tree

. ./common/filter
. ./common/filter.btrfs

_supported_fs btrfs
_require_btrfs_command inspect-internal dump-tree
_require_btrfs_mkfs_feature "raid-stripe-tree"
_require_scratch_dev_pool 4
_require_btrfs_fs_feature "raid_stripe_tree"
_require_btrfs_fs_feature "free_space_tree"
_require_btrfs_free_space_tree
_require_btrfs_no_compress

test $(_get_page_size) -eq 4096 || _notrun "this tests requires 4k pagesize"

test_8k_new_stripe()
{
	local profile=$1
	local ndevs=$2

	_scratch_dev_pool_get $ndevs

	echo "==== Testing $profile ===="
	_scratch_pool_mkfs -d $profile -m $profile -O raid-stripe-tree
	_scratch_mount

	# Fill the first stripe up to 64k - 4k
	$XFS_IO_PROG -fc "pwrite 0 60k" -c fsync "$SCRATCH_MNT/bar" | _filter_xfs_io

	# The actual 8k write
	$XFS_IO_PROG -fc "pwrite 0 8k" "$SCRATCH_MNT/foo" | _filter_xfs_io

	_scratch_cycle_mount
	md5sum "$SCRATCH_MNT/foo" | _filter_scratch

	_scratch_unmount

	$BTRFS_UTIL_PROG inspect-internal dump-tree -t raid_stripe $SCRATCH_DEV_POOL |\
		_filter_stripe_tree

	_scratch_dev_pool_put
}

echo "= Test 8k write to a new file so that 4k start a new stripe ="
test_8k_new_stripe raid0 2
test_8k_new_stripe raid1 2
test_8k_new_stripe raid10 4

# success, all done
status=0
exit
