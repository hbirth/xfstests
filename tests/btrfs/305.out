QA output created by 305
= Test 8k write to a new file so that 4k start a new stripe =
==== Testing raid0 ====
wrote 61440/61440 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 8192/8192 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
9d3940adb41dd525e008a847e01b15f4  SCRATCH_MNT/foo

raid stripe tree key (RAID_STRIPE_TREE ROOT_ITEM 0)
leaf XXXXXXXXX items X free space XXXXX generation X owner RAID_STRIPE_TREE
leaf XXXXXXXXX flags 0x1(WRITTEN) backref revision 1
fs uuid <UUID>
chunk uuid <UUID>
	item 0 key (XXXXXX RAID_STRIPE 61440) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 1 physical XXXXXXXXX
	item 1 key (XXXXXX RAID_STRIPE 4096) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 1 physical XXXXXXXXX
	item 2 key (XXXXXX RAID_STRIPE 4096) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 2 physical XXXXXXXXX
total bytes XXXXXXXX
bytes used XXXXXX
uuid <UUID>
==== Testing raid1 ====
wrote 61440/61440 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 8192/8192 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
9d3940adb41dd525e008a847e01b15f4  SCRATCH_MNT/foo

raid stripe tree key (RAID_STRIPE_TREE ROOT_ITEM 0)
leaf XXXXXXXXX items X free space XXXXX generation X owner RAID_STRIPE_TREE
leaf XXXXXXXXX flags 0x1(WRITTEN) backref revision 1
fs uuid <UUID>
chunk uuid <UUID>
	item 0 key (XXXXXX RAID_STRIPE 61440) itemoff XXXXX itemsize 40
			encoding: RAID1
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 1 key (XXXXXX RAID_STRIPE 8192) itemoff XXXXX itemsize 40
			encoding: RAID1
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
total bytes XXXXXXXX
bytes used XXXXXX
uuid <UUID>
==== Testing raid10 ====
wrote 61440/61440 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 8192/8192 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
9d3940adb41dd525e008a847e01b15f4  SCRATCH_MNT/foo

raid stripe tree key (RAID_STRIPE_TREE ROOT_ITEM 0)
leaf XXXXXXXXX items X free space XXXXX generation X owner RAID_STRIPE_TREE
leaf XXXXXXXXX flags 0x1(WRITTEN) backref revision 1
fs uuid <UUID>
chunk uuid <UUID>
	item 0 key (XXXXXX RAID_STRIPE 61440) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 1 key (XXXXXX RAID_STRIPE 4096) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 2 key (XXXXXX RAID_STRIPE 4096) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 3 physical XXXXXXXXX
			stripe 1 devid 4 physical XXXXXXXXX
total bytes XXXXXXXX
bytes used XXXXXX
uuid <UUID>
